src/dialogs/symlink.tsx

```diff
@@ -0,0 +1,244 @@
+
+import { useFilesContext, type FolderFileInfo } from '../common.ts';
+import { get_owner_candidates } from '../ownership.tsx';
+
+const _ = cockpit.gettext;
+
+type symlink_mode = "absolute" | "relative";
+
+function checkLinkName(candidate: string, entries: Record<string, FileInfo>) {
+    if (candidate === "") {
+        return _("Name cannot be empty");
+    } else if (candidate in entries) {
+        if (entries[candidate].type === "dir") {
+            return _("Directory with the same name exists");
+        }
+        return _("File exists");
+    } else {
+        return null;
+    }
+}
+
+const CreateLinkModal = ({ dialogResult, path, selected } : {
+    dialogResult: DialogResult<void>
+    path: string,
+    selected: FolderFileInfo,
+}) => {
+    const [errorMessage, setErrorMessage] = React.useState<string | undefined>();
+    const [symlinkName, setSymlinkName] = React.useState<string>(cockpit.format(_("link to $0"), selected.name));
+    const [symlinkError, setSymlinkError] = React.useState<string | null>(null);
+    const [mode, setMode] = React.useState<symlink_mode>("relative");
+
+    const { cwdInfo } = useFilesContext();
+
+    const createLink = async () => {
+        cockpit.assert(cwdInfo !== null);
+
+        const cmd = ["ln", "--symbolic", "--no-target-directory"];
+        const target = `${path}${selected.name}`;
+        let symlinkPath = symlinkName;
+        let owner = null;
+
+        // for non-absolute paths obtain the absolute path
+        if (!symlinkPath.startsWith('/')) {
+            try {
+                symlinkPath = await cockpit.spawn(["realpath", `${path}${symlinkName}`], { superuser: "try" });
+                symlinkPath = symlinkPath.trim();
+            } catch (exc) {
+                const err = exc as BasicError;
+                setErrorMessage(err.toString());
+                console.warn("unable to obtain realpath", exc);
+                return;
+            }
+        }
+
+        let folder_info = cwdInfo;
+        // If the path is not in the current directory obtain the user/group/mode
+        if (symlinkName.startsWith('/') || symlinkName.startsWith('..')) {
+            try {
+                folder_info = await fsinfo(dirname(symlinkPath), ["user", "group", "mode"], { superuser: "try" });
+            } catch (exc) {
+                const err = exc as BasicError;
+                setErrorMessage(err.toString());
+                console.warn(`fsinfo failed for ${symlinkPath}`, exc);
+                return;
+            }
+        }
+
+        if (mode === "relative") {
+            cmd.push("--relative");
+        }
+
+        cmd.push(target, symlinkPath);
+
+        if (superuser.allowed) {
+            owner = [...get_owner_candidates(folder_info)][0];
+        }
+
+        try {
+            await cockpit.spawn(cmd, {
+                err: "message",
+                directory: path,
+                ...owner && { superuser: "require" }
```

Comment:
I had to stare at this for a bit. It doesn't mean `(...owner) && {..}` but `...(owner && {..})`.

Reply:
that makes hmm no sense, actually this whole thing is interesting right?

```
        if (superuser.allowed) {
            owner = [...get_owner_candidates(folder_info)][0];
        }
```

So basically only adding superuser: "require" when it is allowed, so we could change this to "try" imo.

Reply:
I meant that the code is right, it's just not very obvious with that spacing and the lack of parentheses. If `owner` is `null`, then this resolves to `{...false} === {}`, and otherwise to `{ superuser: "require" }`

Reply:
I.e. this was just thinking aloud, you don't have to change it and this thread can just be resolved.

Reply:
But ...(owner && {}) is but I am not sure if it is more readable?

Reply:
Because this is a common pattern:

```
        setAlerts(prevAlerts => [
            ...prevAlerts, {
                title,
                variant,
                key,
                ...detail && { detail },
                ...actionLinks && { actionLinks },
            }
```

Reply:
's all good :grin: